tests/Unit/NumericalAlgorithms/LinearSolver/Test_DistributedConjugateGradientAlgorithm.cpp

```diff
@@ -0,0 +1,290 @@
+    Parallel::get<0>(reduced) += Parallel::get<0>(current);
+  }
+  return Parallel::new_reduction_msg(reduced);
+}
+
+struct CollectAp;
+
+struct ComputeOperatorAction {
+  template <typename DbTagsList, typename... InboxTags, typename Metavariables,
+            typename ActionList, typename ParallelComponent>
+  static auto apply(db::DataBox<DbTagsList>& box,
+                    tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    const Parallel::ConstGlobalCache<Metavariables>& cache,
+                    const int array_index, const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/) noexcept {
+    const auto& A = gsl::at(operator_matrices, array_index);
+    const auto& p =
+        get<db::add_tag_prefix<LinearSolver::Tags::Operand, VariablesTag>>(box);
+
+    VariablesType Ap{number_of_grid_points * number_of_elements};
+    dgemv_('N', A.rows(), A.columns(), 1, A.data(), A.rows(), p.data(), 1, 0,
+           Ap.data(), 1);
+
+    Parallel::contribute_to_reduction<&sum_variables, CollectAp>(
+        sum_variables_reduction_data{Ap},
+        Parallel::get_parallel_component<ParallelComponent>(cache)[array_index],
+        Parallel::get_parallel_component<ParallelComponent>(cache));
+
+    // Terminate algorithm for now. The reduction will be broadcasted to the
+    // next action which is responsible for restarting the algorithm.
+    return std::tuple<db::DataBox<DbTagsList>&&, bool>(std::move(box), true);
+  }
+};
+
+struct CollectAp {
+  template <
+      typename... DbTags, typename... InboxTags, typename Metavariables,
+      typename ActionList, typename ParallelComponent,
+      Requires<tmpl2::flat_any_v<cpp17::is_same_v<VariablesTag, DbTags>...>> =
+          nullptr>
+  static auto apply(
+      db::DataBox<tmpl::list<DbTags...>>& box,
+      tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      const Parallel::ConstGlobalCache<Metavariables>& cache,
+      const int array_index, const ActionList /*meta*/,
+      const ParallelComponent* const /*component*/,
+      const sum_variables_reduction_data& Ap_global_data) noexcept {
+    // This could be generalized to work on the Variables instead of the
+    // Scalar, but it's only for the purpose of this test.
+    const auto& Ap_global =
+        get<ScalarFieldTag>(Parallel::get<0>(Ap_global_data)).get();
+    DataVector Ap_local{number_of_grid_points};
+    std::copy(
+        Ap_global.begin() + array_index * int(number_of_grid_points),
+        Ap_global.begin() + (array_index + 1) * int(number_of_grid_points),
+        Ap_local.begin());
+    db::mutate<db::add_tag_prefix<
+        LinearSolver::Tags::OperatorAppliedTo,
+        db::add_tag_prefix<LinearSolver::Tags::Operand, ScalarFieldTag>>>(
+        make_not_null(&box), [&Ap_local](auto Ap) noexcept {
+          *Ap = Scalar<DataVector>(Ap_local);
+        });
+    // Proceed with algorithm
+    Parallel::get_parallel_component<ParallelComponent>(cache)[array_index]
```

Comment:
here too can you just not call `perform_algorithm()` when you're done?

Reply:
please see above